runner.py

```diff
@@ -0,0 +1,653 @@
+import argparse
+import importlib.util
+import logging
+import os
+import pathlib
+import shutil
+import signal
+import subprocess
+import sys
+
+import yaml
+import serial
+import pexpect
+import pexpect.fdpexpect
+
+SCRIPT_PATH = pathlib.Path(__file__).absolute().resolve()
+PRTOS_TEST_DIR = SCRIPT_PATH.parents[0]
+PRTOS_PROJECT_DIR = SCRIPT_PATH.parents[1]
+_BOOT_DIR = PRTOS_PROJECT_DIR / "_boot"
+
+
+QEMU_CMD = {
+    'ia32-generic': [
+        ('qemu-system-i386',),
+        ('-hda', str(_BOOT_DIR) + '/phoenix-ia32-generic.disk'),
+        ('-nographic',),
+        ('-monitor', 'none'),
+        ('-serial', 'stdio')
+    ],
+    'riscv64-virt': [
+        ('qemu-system-riscv64',),
+        ('-bios', str(_BOOT_DIR) + '/phoenix-riscv64-virt.bbl'),
+        ('-machine', 'virt'),
+        ('-nographic',),
+        ('-monitor', 'none'),
+        ('-serial', 'stdio')
+    ]
+}
+
+TARGETS = ['ia32-generic']
+
+
+class Color:
+    FAIL = '\033[91m'
+    OK = '\033[92m'
+    IGNORE = '\033[93m'
+    END = '\033[0m'
+
+    @staticmethod
+    def colorify(string, color):
+        return f"{color}{string}{Color.END}"
+
+
+class DeviceRunner:
+    def __init__(self, port):
+        self.port = port
+        self.serial = None
+
+    def run(self, test):
+        try:
+            self.serial = serial.Serial(self.port, baudrate=115200)
+        except serial.SerialException as exc:
+            test.failed()
+            raise exc
+
+        proc = pexpect.fdpexpect.fdspawn(self.serial, encoding='utf-8')
+
+        try:
+            test.handle(proc)
+        finally:
+            self.serial.close()
+
+
+class QemuRunner:
+    def __init__(self, run_cmd):
+        self.run_cmd = run_cmd
+        # Remove existing serial flag and replace it with stdio
+        self.run_cmd = [flag for flag in self.run_cmd if flag[0] != '-serial']
+        self.run_cmd.append(('-serial', 'stdio'))
+        # Flatten list and turn it into string
+        self.run_cmd = " ".join(list(sum(self.run_cmd, ())))
```

Comment:
```suggestion
        # Remove existing serial flag and replace it with stdio
        self.run_cmd = [flag for flag in run_cmd if flag[0] != '-serial']
        self.run_cmd.append(('-serial', 'stdio'))
        # Flatten list and turn it into string
        self.run_cmd = " ".join(sum(self.run_cmd, ()))
```
`list()` is no needed, `sum()` return `tuple`.
Instead of `sum()` you  can use `itertools.chain()`